### DISCUSSION TITLE
Partial application of function parameters

### DISCUSSION THREAD

**Discussion by jamesw201** [vlang/v#11810] [Ideas]
Partial application is a very useful feature in any language. It could complement vlang's current capabilities and allow for more functional application designs.

For instance:
You might partially apply functions with any dependencies they might need at the top of the application so that you can mock every side effect in the application from the main entry file.

Question: is there any way that a workaround could be achieved using existing features of the language? Perhaps using reflection?

**Comment by andrewboldi** [comment]
I apologize if I am mistaken, but when you refer to a partial function application, are you asking about specifying the number of parameters allowed in a function?

**Comment by JalonSolov** [comment]
Are you talking about this?
```
Partial function application refers to calling a multiple parameter function with less than its total number of
parameters. This results in a new function taking the remaining number of parameters.
```
No, V doesn't have anything like that, and V doesn't do reflection (too slow).

You can write a variadic function, a generic function, etc.  You may be able to get what you need with those.

**Comment by jamesw201** [comment]
apologies @cbracketdash I could have been more specific, @JalonSolov has it right with that definition. 

I imagined it might work something like this:
```
fn multi_param(first int, second int) int {
    return first + second
}
partially_applied := multi_param(3)

print(partially_applied(2))
// 5
```

I think it comes into its own where you have a function that accepts an async dependency like an http client:
```
fn get_user_client(http_client http, user_id int) []User {
    data := http_client.get('https://domain.com/users/$user_id')?
    return data
}
```
This can be imported in main.v and partially applied with the `http` dependency and fed into another application component:
```
import net.http
get_user = get_user_client(http)

fn main() {
    send_out_emails(get_user)
}
```
which no longer needs to know about the dependency when it calls our function:
```
users := get_user(user_id)
```
The result is that you can push all side-effect causing dependencies to main.v where they can be easily mocked, so all of your code can be tested from a unit test at the top of the application. I think this would nicely complement the already strong functional foundation that V has.

Either way I appreciate your responding, this project look great and judging by your responses the community is great too so I look forward to using it.

**Comment by mcastorina** [comment]
I like the idea of partials, but automatically creating one for a function that is called with not enough parameters is not explicit enough for my taste. I would prefer either a new keyword, or a module that can be imported.

**Comment by dumblob** [comment]
The use cases of partial application might be in most cases covered with [closures](https://github.com/vlang/v/issues/7740 ) (unfortunately not yet implemented in V).

**Comment by francescortiz** [comment]
I find very explicit and flexible the approach followed by [gleam](https://gleam.run) for partial application because:
- It is flexible about the order in which the function parameters have been defined.
- It is explicit about how many arguments you are leaving unapplied.

V example of gleam's approach:

```v
fn power(a int, b int) int {
    a ^ b
}

fn main() {
    // one underscore for each non-applied parameter
    square_area = power(_, 2)
    bits_permutations = power(2, _)

    println('area of square with side of 3: ${square_area(3)}') // -> 9
    println('max_value for 3 bits: ${bits_permutations(3) - 1}') // -> 7
}
```

So, even if partial application is implemented with closures, this could be nice syntactic sugar for this use case.

**Reply by Wajinn** [reply]
I like this idea.  Of course it would be for team V to chime in on how doable it is or if there is some unacceptable complication that it may cause.
